File: comparison_tests/test_all_scenarios.py
#!/usr/bin/env python3
"""
Comprehensive test suite for MOCAT-MC Python implementation
Tests all major functions and scenarios for MATLAB comparison
"""

import numpy as np
import sys
import os
import time
from datetime import datetime

# Add parent directory to path to import mocat_mc
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from mocat_mc import MOCATMC


class MOCATTestSuite:
    def __init__(self):
        self.mocat = MOCATMC()
        self.results = {}
        
    def run_test(self, test_name, seed, config_modifier, ic_file='2020.mat'):
        """Run a single test scenario"""
        print(f"\n{'='*60}")
        print(f"Running Test: {test_name}")
        print(f"Seed: {seed}, IC File: {ic_file}")
        print(f"{'='*60}")
        
        try:
            # Start timer
            start_time = time.time()
            
            # Setup configuration
            cfg_mc = self.mocat.setup_mc_config(seed, ic_file)
            
            # Apply test-specific modifications
            config_modifier(cfg_mc)
            
            # Get initial stats
            initial_pop = cfg_mc['mat_sats'].shape[0]
            n_time = cfg_mc['n_time']
            dt_days = cfg_mc['dt_days']
            
            # Run simulation
            nS, nD, nN, nB, mat_sats = self.mocat.main_mc(cfg_mc, seed)
            
            # Calculate results
            elapsed_time = time.time() - start_time
            total_objects = nS + nD + nN + nB
            
            # Store results
            result = {
                'test_name': test_name,
                'seed': seed,
                'success': True,
                'initial_pop': initial_pop,
                'final_pop': total_objects,
                'nS': nS,
                'nD': nD,
                'nN': nN,
                'nB': nB,
                'n_time_steps': n_time,
                'dt_days': dt_days,
                'elapsed_time': elapsed_time,
                'satellite_ratio': nS / total_objects if total_objects > 0 else 0,
                'population_change': total_objects - initial_pop,
                'config': {k: v for k, v in cfg_mc.items() if k not in ['mat_sats', 'tsince']},
                'error': None
            }
            
            print(f"Initial Population: {initial_pop}")
            print(f"Final Population: {total_objects} (change: {total_objects - initial_pop:+d})")
            print(f"Final Counts - S: {nS}, D: {nD}, N: {nN}, B: {nB}")
            print(f"Satellite Ratio: {result['satellite_ratio']:.4f}")
            print(f"Execution Time: {elapsed_time:.2f} seconds")
            
        except Exception as e:
            print(f"ERROR in test {test_name}: {str(e)}")
            result = {
                'test_name': test_name,
                'seed': seed,
                'success': False,
                'error': str(e)
            }
            
        self.results[test_name] = result
        return result
    
    # Test configuration functions
    def basic_propagation_config(self, cfg):
        """Test 1: Basic propagation without collisions"""
        cfg['n_time'] = 10
        cfg['skipCollisions'] = 1
        cfg['launch_model'] = 'no_launch'
        cfg['P_frag'] = 0
        
    def collision_detection_config(self, cfg):
        """Test 2: Collision detection"""
        cfg['dt_days'] = 1
        cfg['CUBE_RES'] = 100
        cfg['alph'] = 0.01
        cfg['alph_a'] = 0
        cfg['skipCollisions'] = 0
        cfg['n_time'] = 36  # ~6 months with daily steps
        
    def fragmentation_config(self, cfg):
        """Test 3: Fragmentation/explosion test"""
        cfg['P_frag'] = 1e-5
        cfg['P_frag_cutoff'] = 20
        cfg['max_frag'] = 100
        cfg['n_time'] = 36  # ~6 months
        
    def orbit_control_config(self, cfg):
        """Test 4: Orbit control test"""
        cfg['orbtol'] = 10
        cfg['step_control'] = 5
        cfg['PMD'] = 0.90
        cfg['missionlifetime'] = 5
        
    def no_launch_config(self, cfg):
        """Test 5: No launch baseline"""
        cfg['launch_model'] = 'no_launch'
        cfg['n_time'] = 73  # 1 year
        
    def matsat_launch_config(self, cfg):
        """Test 6: MatSat launch model"""
        cfg['launch_model'] = 'matsat'
        cfg['launchRepeatYrs'] = [2018, 2022]
        cfg['launchRepeatSmooth'] = 0
        cfg['n_time'] = 146  # 2 years
        
    def matsat_smooth_config(self, cfg):
        """Test 7: MatSat with smoothing"""
        cfg['launch_model'] = 'matsat'
        cfg['launchRepeatYrs'] = [2015, 2020]
        cfg['launchRepeatSmooth'] = 1
        cfg['n_time'] = 219  # 3 years
        
    def atmospheric_drag_config(self, cfg):
        """Test 8: Atmospheric drag focus"""
        cfg['altitude_limit_low'] = 200
        cfg['altitude_limit_up'] = 600
        cfg['dt_days'] = 0.5
        cfg['n_time'] = 182  # 3 months with half-day steps
        
    def high_activity_config(self, cfg):
        """Test 9: High activity period"""
        cfg['launch_model'] = 'matsat'
        cfg['launchRepeatYrs'] = [2019, 2022]
        cfg['launchRepeatSmooth'] = 0
        cfg['missionlifetime'] = 5
        cfg['PMD'] = 0.90
        cfg['alph'] = 0.02
        cfg['n_time'] = 365  # 5 years
        
    def extreme_altitude_config(self, cfg):
        """Test 10: Extreme altitude test"""
        cfg['altitude_limit_low'] = 150
        cfg['altitude_limit_up'] = 50000
        cfg['collision_alt_limit'] = 50000
        cfg['n_time'] = 73  # 1 year
        
    def mixed_scenario_config(self, cfg):
        """Test 11: Mixed launch and collision"""
        cfg['launch_model'] = 'matsat'
        cfg['launchRepeatYrs'] = [2018, 2020]
        cfg['skipCollisions'] = 0
        cfg['alph'] = 0.05
        cfg['CUBE_RES'] = 100
        cfg['n_time'] = 146  # 2 years
        
    def full_integration_config(self, cfg):
        """Test 12: Full integration with defaults"""
        # Use all defaults - 1 year simulation
        pass
    
    def run_all_tests(self):
        """Run complete test suite"""
        print("\n" + "="*80)
        print("MOCAT-MC PYTHON IMPLEMENTATION TEST SUITE")
        print(f"Started at: {datetime.now()}")
        print("="*80)
        
        # Define all test scenarios
        test_scenarios = [
            # Basic functionality tests
            ("Basic Propagation", 42, self.basic_propagation_config),
            ("Collision Detection", 123, self.collision_detection_config),
            ("Fragmentation Test", 999, self.fragmentation_config),
            ("Orbit Control", 777, self.orbit_control_config),
            
            # Launch model tests
            ("No Launch Baseline", 100, self.no_launch_config),
            ("MatSat Launch", 200, self.matsat_launch_config),
            ("MatSat Smoothed", 300, self.matsat_smooth_config),
            
            # Environmental tests
            ("Atmospheric Drag", 333, self.atmospheric_drag_config),
            ("Extreme Altitudes", 1100, self.extreme_altitude_config),
            
            # Complex scenarios
            ("High Activity Period", 500, self.high_activity_config),
            ("Mixed Launch/Collision", 600, self.mixed_scenario_config),
            ("Full Integration", 1, self.full_integration_config),
        ]
        
        # Run each test
        for test_name, seed, config_func in test_scenarios:
            self.run_test(test_name, seed, config_func)
        
        # Print summary
        self.print_summary()
        
    def print_summary(self):
        """Print test results summary"""
        print("\n" + "="*80)
        print("TEST RESULTS SUMMARY")
        print("="*80)
        
        successful = sum(1 for r in self.results.values() if r.get('success', False))
        total = len(self.results)
        
        print(f"\nTotal Tests: {total}")
        print(f"Successful: {successful}")
        print(f"Failed: {total - successful}")
        
        print("\nDetailed Results:")
        print(f"{'Test Name':<25} {'Status':<10} {'Initial':<8} {'Final':<8} {'S':<6} {'D':<6} {'N':<6} {'B':<6} {'Time(s)':<8}")
        print("-" * 100)
        
        for test_name, result in self.results.items():
            if result.get('success', False):
                print(f"{test_name:<25} {'PASS':<10} "
                      f"{result['initial_pop']:<8} {result['final_pop']:<8} "
                      f"{result['nS']:<6} {result['nD']:<6} "
                      f"{result['nN']:<6} {result['nB']:<6} "
                      f"{result['elapsed_time']:<8.2f}")
            else:
                print(f"{test_name:<25} {'FAIL':<10} Error: {result.get('error', 'Unknown')}")
        
        # Save results to file
        self.save_results()
        
    def save_results(self):
        """Save results to file for MATLAB comparison"""
        import json
        
        filename = f"python_test_results_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
        with open(filename, 'w') as f:
            json.dump(self.results, f, indent=2, default=str)
        print(f"\nResults saved to: {filename}")


if __name__ == "__main__":
    # Run test suite
    test_suite = MOCATTestSuite()
    test_suite.run_all_tests()

File: MATLAB_implementation/tests/test_all_scenarios.m
%% MOCAT-MC MATLAB Test Suite
% Comprehensive test scenarios matching Python implementation
% Run all tests and save results for comparison

function test_all_scenarios()
    % Add paths
    addpath(genpath('../supporting_data/')); 
    addpath(genpath('../supporting_functions'));
    
    fprintf('\n%s\n', repmat('=', 1, 80));
    fprintf('MOCAT-MC MATLAB IMPLEMENTATION TEST SUITE\n');
    fprintf('Started at: %s\n', datestr(now));
    fprintf('%s\n', repmat('=', 1, 80));
    
    % Initialize results structure
    results = {};
    
    % Define all test scenarios matching Python
    test_scenarios = {
        % Test Name,              Seed, Config Function
        'Basic Propagation',      42,   @basic_propagation_config;
        'Collision Detection',    123,  @collision_detection_config;
        'Fragmentation Test',     999,  @fragmentation_config;
        'Orbit Control',          777,  @orbit_control_config;
        'No Launch Baseline',     100,  @no_launch_config;
        'MatSat Launch',          200,  @matsat_launch_config;
        'MatSat Smoothed',        300,  @matsat_smooth_config;
        'Atmospheric Drag',       333,  @atmospheric_drag_config;
        'Extreme Altitudes',      1100, @extreme_altitude_config;
        'High Activity Period',   500,  @high_activity_config;
        'Mixed Launch/Collision', 600,  @mixed_scenario_config;
        'Full Integration',       1,    @full_integration_config;
    };
    
    % Run each test
    for i = 1:size(test_scenarios, 1)
        test_name = test_scenarios{i, 1};
        seed = test_scenarios{i, 2};
        config_func = test_scenarios{i, 3};
        
        result = run_test(test_name, seed, config_func);
        results{end+1} = result;
    end
    
    % Print summary
    print_summary(results);
    
    % Save results for Python comparison
    save_results(results);
end

%% Individual test runner
function result = run_test(test_name, seed, config_func)
    fprintf('\n%s\n', repmat('=', 1, 60));
    fprintf('Running Test: %s\n', test_name);
    fprintf('Seed: %d\n', seed);
    fprintf('%s\n', repmat('=', 1, 60));
    
    try
        % Start timer
        tic;
        
        % Setup configuration
        cfg_mc = setup_MCconfig(seed, '2020.mat');
        
        % Apply test-specific modifications
        cfg_mc = config_func(cfg_mc);
        
        % Get initial stats
        initial_pop = size(cfg_mc.mat_sats, 1);
        n_time = cfg_mc.n_time;
        dt_days = cfg_mc.dt_days;
        
        % Run simulation
        [nS, nD, nN, nB, mat_sats] = main_mc(cfg_mc, seed);
        
        % Calculate results
        elapsed_time = toc;
        total_objects = nS + nD + nN + nB;
        
        % Store results
        result = struct();
        result.test_name = test_name;
        result.seed = seed;
        result.success = true;
        result.initial_pop = initial_pop;
        result.final_pop = total_objects;
        result.nS = nS;
        result.nD = nD;
        result.nN = nN;
        result.nB = nB;
        result.n_time_steps = n_time;
        result.dt_days = dt_days;
        result.elapsed_time = elapsed_time;
        result.satellite_ratio = nS / total_objects;
        result.population_change = total_objects - initial_pop;
        result.error = '';
        
        % Sample orbital elements for comparison (first 10 objects)
        n_sample = min(10, size(mat_sats, 1));
        result.sample_oe = mat_sats(1:n_sample, 1:6);
        result.sample_rv = mat_sats(1:n_sample, 17:22);
        
        fprintf('Initial Population: %d\n', initial_pop);
        fprintf('Final Population: %d (change: %+d)\n', total_objects, total_objects - initial_pop);
        fprintf('Final Counts - S: %d, D: %d, N: %d, B: %d\n', nS, nD, nN, nB);
        fprintf('Satellite Ratio: %.4f\n', result.satellite_ratio);
        fprintf('Execution Time: %.2f seconds\n', elapsed_time);
        
    catch ME
        fprintf('ERROR in test %s: %s\n', test_name, ME.message);
        result = struct();
        result.test_name = test_name;
        result.seed = seed;
        result.success = false;
        result.error = ME.message;
    end
end

%% Test configuration functions (matching Python exactly)

function cfg = basic_propagation_config(cfg)
    % Test 1: Basic propagation without collisions
    cfg.n_time = 10;
    cfg.skipCollisions = 1;
    cfg.launch_model = 'no_launch';
    cfg.P_frag = 0;
end

function cfg = collision_detection_config(cfg)
    % Test 2: Collision detection
    cfg.dt_days = 1;
    cfg.DeltaT = cfg.dt_days * cfg.DAY2MIN;
    cfg.tsince = 0:cfg.DeltaT:(36*cfg.DeltaT);  % ~6 months
    cfg.n_time = length(cfg.tsince);
    cfg.CUBE_RES = 100;
    cfg.alph = 0.01;
    cfg.alph_a = 0;
    cfg.skipCollisions = 0;
end

function cfg = fragmentation_config(cfg)
    % Test 3: Fragmentation/explosion test
    cfg.P_frag = 1e-5;
    cfg.P_frag_cutoff = 20;
    cfg.max_frag = 100;
    cfg.n_time = 36;  % ~6 months
end

function cfg = orbit_control_config(cfg)
    % Test 4: Orbit control test
    cfg.orbtol = 10;
    cfg.step_control = 5;
    cfg.PMD = 0.90;
    cfg.missionlifetime = 5;
end

function cfg = no_launch_config(cfg)
    % Test 5: No launch baseline
    cfg.launch_model = 'no_launch';
    cfg.n_time = 73;  % 1 year
end

function cfg = matsat_launch_config(cfg)
    % Test 6: MatSat launch model
    cfg.launch_model = 'matsat';
    cfg.launchRepeatYrs = [2018, 2022];
    cfg.launchRepeatSmooth = 0;
    cfg.n_time = 146;  % 2 years
end

function cfg = matsat_smooth_config(cfg)
    % Test 7: MatSat with smoothing
    cfg.launch_model = 'matsat';
    cfg.launchRepeatYrs = [2015, 2020];
    cfg.launchRepeatSmooth = 1;
    cfg.n_time = 219;  % 3 years
end

function cfg = atmospheric_drag_config(cfg)
    % Test 8: Atmospheric drag focus
    cfg.altitude_limit_low = 200;
    cfg.altitude_limit_up = 600;
    cfg.dt_days = 0.5;
    cfg.DeltaT = cfg.dt_days * cfg.DAY2MIN;
    cfg.tsince = 0:cfg.DeltaT:(182*cfg.DeltaT);  % 3 months
    cfg.n_time = length(cfg.tsince);
end

function cfg = extreme_altitude_config(cfg)
    % Test 10: Extreme altitude test
    cfg.altitude_limit_low = 150;
    cfg.altitude_limit_up = 50000;
    cfg.collision_alt_limit = 50000;
    cfg.n_time = 73;  % 1 year
end

function cfg = high_activity_config(cfg)
    % Test 9: High activity period
    cfg.launch_model = 'matsat';
    cfg.launchRepeatYrs = [2019, 2022];
    cfg.launchRepeatSmooth = 0;
    cfg.missionlifetime = 5;
    cfg.PMD = 0.90;
    cfg.alph = 0.02;
    cfg.n_time = 365;  % 5 years
end

function cfg = mixed_scenario_config(cfg)
    % Test 11: Mixed launch and collision
    cfg.launch_model = 'matsat';
    cfg.launchRepeatYrs = [2018, 2020];
    cfg.skipCollisions = 0;
    cfg.alph = 0.05;
    cfg.CUBE_RES = 100;
    cfg.n_time = 146;  % 2 years
end

function cfg = full_integration_config(cfg)
    % Test 12: Full integration with defaults
    % Use all defaults - 1 year simulation
end

%% Summary printer
function print_summary(results)
    fprintf('\n%s\n', repmat('=', 1, 80));
    fprintf('TEST RESULTS SUMMARY\n');
    fprintf('%s\n', repmat('=', 1, 80));
    
    successful = sum(cellfun(@(x) x.success, results));
    total = length(results);
    
    fprintf('\nTotal Tests: %d\n', total);
    fprintf('Successful: %d\n', successful);
    fprintf('Failed: %d\n', total - successful);
    
    fprintf('\nDetailed Results:\n');
    fprintf('%-25s %-10s %-8s %-8s %-6s %-6s %-6s %-6s %-8s\n', ...
        'Test Name', 'Status', 'Initial', 'Final', 'S', 'D', 'N', 'B', 'Time(s)');
    fprintf('%s\n', repmat('-', 1, 100));
    
    for i = 1:length(results)
        r = results{i};
        if r.success
            fprintf('%-25s %-10s %-8d %-8d %-6d %-6d %-6d %-6d %-8.2f\n', ...
                r.test_name, 'PASS', r.initial_pop, r.final_pop, ...
                r.nS, r.nD, r.nN, r.nB, r.elapsed_time);
        else
            fprintf('%-25s %-10s Error: %s\n', r.test_name, 'FAIL', r.error);
        end
    end
end

%% Save results for comparison
function save_results(results)
    % Save as MAT file for easy Python loading
    timestamp = datestr(now, 'yyyymmdd_HHMMSS');
    filename = sprintf('matlab_test_results_%s.mat', timestamp);
    
    % Convert cell array to structure array for better compatibility
    results_struct = [];
    for i = 1:length(results)
        results_struct = [results_struct, results{i}];
    end
    
    save(filename, 'results_struct', '-v7');
    fprintf('\nResults saved to: %s\n', filename);
    
    % Also save as CSV for easy viewing
    csv_filename = sprintf('matlab_test_summary_%s.csv', timestamp);
    fid = fopen(csv_filename, 'w');
    fprintf(fid, 'Test_Name,Status,Seed,Initial_Pop,Final_Pop,nS,nD,nN,nB,Time_s\n');
    
    for i = 1:length(results)
        r = results{i};
        if r.success
            fprintf(fid, '%s,PASS,%d,%d,%d,%d,%d,%d,%d,%.2f\n', ...
                r.test_name, r.seed, r.initial_pop, r.final_pop, ...
                r.nS, r.nD, r.nN, r.nB, r.elapsed_time);
        else
            fprintf(fid, '%s,FAIL,%d,,,,,,,\n', r.test_name, r.seed);
        end
    end
    fclose(fid);
    fprintf('Summary saved to: %s\n', csv_filename);
end